[JIT] Cleanup special handling of Containers, allowing custom forwards

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29495 add support for _modules, reducing special casing of nn.Sequential
* **#28988 [JIT] Cleanup special handling of Containers, allowing custom forwards**
* #28987 Make nn.Sequential iterable

Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?

Differential Revision: [D18402821](https://our.internmc.facebook.com/intern/diff/D18402821)

Commit:
[JIT] Cleanup special handling of Containers, allowing custom forwards

[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"




[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"




[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

[ghstack-poisoned]

driazati commented on torch/jit/_recursive.py:
```diff
@@ -68,9 +68,9 @@ def infer_raw_concrete_type(nn_module):
     """
     concrete_type = torch._C.ConcreteModuleType()
     concrete_type.add_pyclass(type(nn_module))
-    if isinstance(nn_module, (torch.nn.ModuleDict, torch.jit._ConstModuleDict)):
```
I think this will break in a case like this, due to #29092

```python
class S(Sequential):
   ....

class M(Module):
    def __init__(self):
        super().__init__()
        self.x = torch.jit.script(S())

torch.jit.script(M())
```

eellison replied:
Looking into this example, however https://github.com/pytorch/pytorch/issues/29092 should be fixed on its own right as well.

eellison replied:
This works:
```
        class Inner(torch.nn.Module):
            def forward(self, x):
                return x + 10

        class CustomSequential(nn.Sequential):
            def __init__(self):
                super(CustomSequential, self).__init__(
                    nn.ReLU(), Inner())

            def forward(self, x):
                x = x + 3
                for mod in self:
                    x = mod(x)
                return x - 5

        class M(nn.Module):
            def __init__(self):
                super().__init__()
                self.x = torch.jit.script(CustomSequential())

            def forward(self, input):
                return self.x.forward(input)

        self.checkModule(M(), (torch.tensor(.5),))
```

driazati commented on torch/jit/__init__.py:
```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
+        # define magic methods here as a shim to the correct attribute.
+        def forward_magic_method(self, method_name, *args, **kwargs):
+            self_method = getattr(self, method_name)
+            if self_method == getattr(RecursiveScriptModule, method_name):
```
Won't this break `__dir__` calls for normal `ScriptModule`s?

eellison replied:
What are you worried about breaking ? Shouldn't the behavior inherit from `ModuleList`, which does the following:

```
        keys = super(ModuleList, self).__dir__()
        keys = [key for key in keys if not key.isdigit()]
        return keys
```

(and which this currently calls into)

driazati commented on torch/jit/__init__.py:
```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
+        def __iter__(self):
+            return self.forward_magic_method("__iter__")
+
+        def __getitem__(self, idx):
```
Defining the methods this way instead of something like `_ConstModuleList` before will make the error messages confusing for something like re-assigning a module since it would say it has no `__setitem__` rather than saying something about how compiled `ModuleList`s are constant

eellison replied:
Could you give an example of when you think the error messages would be confusing ? I included an error message above when this code path gets called: "object is not iterable"

suo commented on torch/jit/_recursive.py:
```diff
@@ -403,8 +384,58 @@ def init_fn(script_module):
+    # In order to continue to expose module container functions to python,
+    # we add on the methods that we had previously exposed in the previous
+    # version of this api.
+    if isinstance(nn_module, (ModuleList, Sequential)):
```
It would be awesome of we could eliminate any special handling for ModuleContainers here at all. The generic mechanism we want here is "make sure this Python method on `nn_module` shows up on the `script_module`". One way to do this it to have an internal decorator called like `@_copy_to_script` (we need a better name), which specifies that a given method should be copied to the new Python ScriptModule. These decorators can go on all the container modules, and `ScriptModule` creation can just process the decorators and do the copy without encoding special-case policies.

eellison replied:
These are special-cased methods that exist only so that we continue to expose the same api for these built-in containers as we did previously. I don't know if it makes sense to generalize it. Another possibility we can at least consider is to break the api for the constant modules. We'd only be breaking it for their usage as scripted modules within python, not within torchscript.

eellison commented on torch/jit/_recursive.py:
```diff
@@ -465,7 +496,8 @@ def infer_methods_to_compile(nn_module):
     if hasattr(nn_module, 'forward'):
-        if getattr(nn_module.forward, "__func__", None) == torch.nn.Module.forward:
+        forward_func = getattr(nn_module.forward, "__func__", None)
+        if forward_func == torch.nn.Module.forward or forward_func == Sequential.forward:
```
ModuleList and ModuleDict don't have a forward, it is already covered by `forward_func == torch.nn.Module.forward` case

suo commented on torch/jit/__init__.py:
```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
                 "For purely script modules use my_script_module.save(<filename>) instead.")
 
+        # Python magic methods do lookups on an object's class type, instead of looking up
+        # the desugared attribute. In order to continue to expose the magic methods
```
nit: I was confused by the word "desguared" here. I think it would be clearer to talk about methods defined on classes vs. methods defined on instances.

eellison commented on test/test_jit.py:
```diff
@@ -8629,6 +8677,9 @@ def forward(self, v):
                 v = sub(v)
             self.assertEqual(o, v)
 
+            with self.assertRaisesRegex(Exception, "object is not iterable"):
```
I don't know, i just put this test where the rest of the functionality is being exposed. I suspect it's not needed.

suo commented on torch/jit/__init__.py:
```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
+        # the desugared attribute. In order to continue to expose the magic methods
+        # of builtin-containers (ModuleList, Sequential, ModuleDict) to python we
+        # define magic methods here as a shim to the correct attribute.
+        def forward_magic_method(self, method_name, *args, **kwargs):
```
It's a little unfortunate that we are littering these magic methods for `dir()` and `inspect` to pick up. One alternative is for `create_script_module` to generate a new type that inherits from `RecursiveScriptModule` and define all `@_copy_to_script` methods on it. Can we try that?

suo replied:
I think it's preferable to separate mechanism from policy, and have the `create_script_module` only know about the mechanism. It's better that the decorators visible I think; if we defined new generic container/sequence classes, or change the magic methods we define on existing ones, we shouldn't expect implementors to know that they need to go to `_recursive` to change. We can consult with the NY team maybe?

As for whether we will expose this to the user—I don't see anything wrong with making `_copy_to_script` public (it's useful and non-duplicative) but we should be careful about increasing the API surface area so I think in this PR we keep it private.

eellison replied:
Separately - what exactly were you envisioning with the `generate a new type` comment ?

suo replied:
> I also think it's best to avoid the pattern of inserting code that only has semantic meaning to the JIT team in the rest of pytorch when we can; in past we've done so e.g. in LSTM with __overload__ and @ignore and it hasn't been great.

I think this is a judgment call. `__overload__` and `@ignore` complicate up the code in parts that need to be readable, this decorator does not. I think it's fine in this case.

As for the "generate a new type" thing, @zdevito convinced me that it's a bad idea. I think we can just use `forward_magic_method` but we need to fix the implementation so that `__dir__` works. Also, it would be easier I think if `forward_magic_method` were called in a loop over a list of names, like we do for copying methods from `RecursiveScriptModule` to `ScriptModule`.

eellison replied:
I don't know how the "loop over a list of names" would work. The method needs to be defined on the class, so copying these as attributes will not work.

eellison replied:
Okay, updated to use wrapper. adding pytorch eager ppl to review. cc @zou3519 @gchanan

suo commented on torch/jit/__init__.py:
```diff
@@ -1731,6 +1731,31 @@ def __getstate__(self):
+        def __len__(self):
+            return self.forward_magic_method("__len__")
+
+        def __dir__(self):
```
`nn.Module` implements `__dir__`, so I think it breaks for regular modules.

eellison replied:
So in the case of `__dir__` we should always `@_copy_to_script` so to speak ? or what should be done here

suo replied:
Since `RecursiveScriptModule` inherits from `nn.Module`, `__dir__` is always defined on `RecursiveScriptModule`s. This makes `forward_magic_method` problematic, because it takes over the default behavior. I think the "generate a new type" strategy would work to fix this, since it would just further override `__dir__`.

suo reviewed (changes requested):
Great change, this simplifies a lot of things! I had a few suggestions to help make it more generic.

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

eellison commented:
@suo as far as I can tell this only has python 2 errors... which maybe doesn't matter anymore ? Can i get another review.

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

suo commented on torch/jit/_recursive.py:
```diff
@@ -456,6 +477,9 @@ def check_module_initialized(mod):
         raise RuntimeError("'{}' has not been initialized, did you forget to call 'super()'?"
                            .format(type(mod).__name__))
 
+def get_unbound_method_func(unbound_method):
```
This is introduced but unused.

suo reviewed (approved):
lgtm, thanks for this!

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?


[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?

Differential Revision: [D18402821](https://our.internmc.facebook.com/intern/diff/D18402821)

[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?

Differential Revision: [D18402821](https://our.internmc.facebook.com/intern/diff/D18402821)

[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?

Differential Revision: [D18402821](https://our.internmc.facebook.com/intern/diff/D18402821)

[ghstack-poisoned]

Commit:
Update on "[JIT] Cleanup special handling of Containers, allowing custom forwards"


Make ModuleList, Sequential, ModuleDict go through the same pathway as other modules, cleaning up a bunch of code and allowing them to define custom forwards and other methods.

EDIT: Previously, we would ignore an nn.Sequential attribute if it was not in `__constants__` ("did you forget to add it to Constants"). This PR scripts it even if it is not in `__constants__`. Is that what we want?

Differential Revision: [D18402821](https://our.internmc.facebook.com/intern/diff/D18402821)

[ghstack-poisoned]
